### ISSUE TITLE
coveryourtracks.eff.org says randomized by first party domain when using spoof

### ISSUE THREAD

**Issue by bm92aWJldnJvcw** [kkapsner/CanvasBlocker#571]
## Description

I just checked on coveryourtracks.eff.org and it says "randomized by first party domain" for canvas, webgl and audiocontext when the spoof (extension) is ON. when it is OFF, it shows the correct hash value. So I'm thinking if it is detecting our spoof? 
## Expected Behavior

coveryourtracks.eff.org should show the random (spoof) value instead of "randomized by first party domain" response for hashes even when CanvasBlocker is ON.
## Current Behavior

coveryourtracks.eff.org shows "randomized by first party domain" response instead of hashes when using CanvasBlocker. 

## Steps to Reproduce (for bugs)

1. Install CanvasBlocker. 
2. Set the preset to Stealth mode.
3. Visit coveryourtracks.eff.org, in result check canvas, webgl and audiocontext response.

## Context

for me the issue is audiocontext hash spoof should not be detectable. 
## Your Environment

* CanvasBlocker Version used: 1.6.1
* Firefox version incl. 32- or 64-bit: 64-bit
* Operating System and version (desktop or mobile): Windows 10
* Installed addons: none other.

**Comment by Thorin-Oakenpants** [comment]
AFAIK cover your tracks compares two readings from different eTLD+1's, and since the randomizing is per eTLD+1, they differ - it's not unmasking the "stealth" randomizing solely via 1st party

That said, randomizing can be detected from known results, no third parties required - see [this](https://arkenfox.github.io/TZP/tests/canvasspoof.html) for canvas - a red x means you lied

**Comment by bm92aWJldnJvcw** [comment]
Oh, now I get it. Thanks for the clarification. In that case can turning ON the "Share persistent randomness between domains" work? I checked it doesn't seem to work still. I've seen amzn doing similar fingerprinting or atleast it fetches fingerprint few times in signup.

**Comment by Thorin-Oakenpants** [comment]
> In that case can turning ON the "Share persistent...

**_NO_**

Randomizing has two key purposes - 1. to hide the real value, because that real value is high entropy and would make your fingerprint easier to link your traffic . In this regard you are safe. CB will not leak your real value. And 2. the random value is a poison pill - i.e scripts that don't check for randomness, swallow the fake data and can't link your traffic if they use it. Most scripts are naive (they get sucked in), but that will probably change over time

At best, scripts can detect that you are lying, and instead of getting sucked in, record a static value for canvas - "noise detected", and thus the randomness is negated, but the real value is not compromised. Your real value is very entropy. Getting assigned a value of "random bullshit" is not as high entropy (most likely) as the number of people randomizing canvas grows - e.g. all RFP users, all CB users, all Brave users and so on.

The bad news is, that smart scripts can fingerprint the randomness charactersistics. Brave has a unique random canvas metric (besides the fact that you can tell it's chromium and actually Brave, as well as it;s shield level), and RFP randomizes per execution and also have a very unique random FP - see [here](https://arkenfox.github.io/TZP/tests/canvasrfp.html) <-- turn on RFP and reload the page. Now allow a site exception and change RFP to not use cache (so it;s random per execution) and load it again, and I can tell its not RFP

And [here](https://arkenfox.github.io/TZP/tests/canvasnoise.html) is an example of detecting some randomness characteristics - I should really finish that test, but you get the idea - e.g. RFP randomizes 4 channels and every single pixel ... CB does soemthing different, etc

long story short - excluding other linkifying issues such as IP addresses - .. on client side JS metrics, when randomizing, you really need them to be per eLTD+1 - otherwise you are giving a free stable probably unique FP to track you across your entire session

hope I made sense, and don't sweat it ... most scripts, if they run, are dumb and swallow the poison pill

**Comment by bm92aWJldnJvcw** [comment]
I realised that recently. Unfortunately, things I do has these smart scripts lol. They may take the poison pill only to make the session flagged F. so I prefer keeping same (real) canvas and webgl with enough randomizing of other values i.e. audiocontext and domrect etc. to be a different machine on each run, it helps defeat fingerprintjs.com along with being in the mass (I think). thanks this issue may be closed now.

**Comment by Thorin-Oakenpants** [comment]
> with enough randomizing of other values i.e. audiocontext and domrect 

you do realize that these can all be detected as well, right?

**Comment by bm92aWJldnJvcw** [comment]
> > with enough randomizing of other values i.e. audiocontext and domrect
> 
> you do realize that these can all be detected as well, right?

Yeah. what can we do :') antidetect softwares are the only way to get genuine everything.

**Comment by kkapsner** [comment]
> Oh, now I get it. Thanks for the clarification. In that case can turning ON the "Share persistent randomness between domains" work? I checked it doesn't seem to work still. I've seen amzn doing similar fingerprinting or atleast it fetches fingerprint few times in signup.

"Share persistent randomness between domains" works for me. BUT you should not use it if you are concerned about tracking/privacy - you are getting a bit fat warning about that when you try to enable it.

> thanks this issue may be closed now.

OK - will do. To be honest there is very little I can do about the detection. CB does stuff and everything that is done can be detected somehow. What I plan to do is to add an option to use the RFP canvas pattern.

**Comment by Thorin-Oakenpants** [comment]
> and everything that is done can be detected somehow

somehow? it's [witchcraft](https://www.mit.edu/afs.new/sipb/user/ayshames/Python/WITCH.PYTHON), also known as mathematics :)

> What I plan to do is to add an option to use the RFP canvas pattern

But will it pass my test? And if it does, then it's the same as not allowing an RFP site exception, in other words a bit pointless (because RFP must be enabled to pass the test). Granted, scripts aren't going to do what I do

**Comment by kkapsner** [comment]
> But will it pass my test? And if it does, then it's the same as not allowing an RFP site exception, in other words a bit pointless (because RFP must be enabled to pass the test). Granted, scripts aren't going to do what I do

Probably not as I will only do the Canvas part (that repeating colour pattern). And yes - I think it will confuse some scripts (for a time). Personally I will not use it.
